### DISCUSSION TITLE
Adressing specific instances of templates in UPPAAL-C

### DISCUSSION THREAD

**Discussion by AstridHornBrorholt** [UPPAALModelChecker/UPPAAL-Meta#158] [Q&A] [answered]
In the requirements specification language, I can address variables of specific instances a template like 

    simulate[<=20;1] { Ball(0).p, Ball(1).p, Ball(2).p }

![image](https://user-images.githubusercontent.com/11016262/197341668-49c6a534-f324-449b-b4cd-f495c78bf59f.png)

If I need to access these same variables in a function in my declarations, how could I do this? The same `identifier(id).identifier` pattern does not work

![image](https://user-images.githubusercontent.com/11016262/197341610-215ab06c-72c6-4fea-8826-fcc772c9dd28.png)

I have a Ball template that has hybrid clocks v and p. This ball template has multiple instances in the system, controlled by a parameter `ball_id id` provided to the template. 

![image](https://user-images.githubusercontent.com/11016262/197341469-9ab457fd-26ad-4c6a-a998-1d461add938d.png)

The id type is defined as:

     const int NUMBER_OF_BALLS = 3;
     typedef int[0, NUMBER_OF_BALLS - 1] ball_id;

Which is implicitly(?) used in this simple system definition:

    system Ball, LearnerPlayer;

**Comment by thorulf4** [comment, marked as answer]
This is not really possible as template variables are not yet declared in the global declarations.

The simplest work around is probably moving your template state into global declarations. With something like: 

```
hybrid clock v[ball_id];
hybrid clock p[ball_id];
```
Then Ball instantiations can use their id to get their own data.

**Comment by AstridHornBrorholt** [comment]
Thank you, that's a good tip. I tried this and everything works.
